mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala

```diff
@@ -920,6 +924,14 @@ object ALS extends DefaultParamsReadable[ALS] with Logging {
     require(intermediateRDDStorageLevel != StorageLevel.NONE,
       "ALS is not designed to run without persisting intermediate RDDs.")
 
+    // Indeterminate rating RDD causes inconsistent in/out blocks in case of rerun.
+    // It can cause runtime error when matching in/out user/item blocks.
+    if (ratings.outputDeterministicLevel == DeterministicLevel.INDETERMINATE) {
+      throw new IllegalArgumentException("The output of rating RDD can not be indeterminate. " +
+        "If your training data has indeterminate RDD computations, like `randomSplit` or `sample`" +
+        ", please checkpoint the training data before running ALS.")
```

Comment:
Or cache, right?

Reply:
Only cache doesn't fix this.

We use cache, but in case of cached blocks are lost, it still triggers rerun of the training RDD tasks.

Reply:
I think this is going to break a lot of user code, although you are strictly speaking quite correct. If I do some train/test split and even cache() the results, which is the usual practice, it will fail now right? because the cached result is not considered deterministic. Does checkpointing change the determinism level? what if you lose the checkpoint?

It's not a wild idea. But this problem exists everywhere in theory in Spark. Any test/train split basically has this problem. I hesitate to enforce this so strictly everywhere? how about a warning?

Reply:
This kind of failure only happens when you lost blocks of training data. That can be cached blocks, or map outputs. It is not definitely going to fail in all cases. But it is more likely, when fitting ALS on big amount data. This is hard to reproduce in unit test or small dataset.

Checkpointing changes RDD output to deterministic:

https://github.com/apache/spark/blob/c610de69526d65f3b679cfd81ab7e1a5791ff37f/core/src/main/scala/org/apache/spark/rdd/RDD.scala#L1919-L1925

I think checkpointed RDD can't be rerun like cache or normal RDD, because checkpointing will clean up RDD lineage. When you lose checkpoint, the job should be failed.

Yeah, a warning sounds good. I also added notes to documents, hopefully this can notify users about this issue.

Reply:
Yes, I meant that throwing an exception is the thing that would definitely cause failures, when it currently happens to work fine in practice in almost all cases.